src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,403 @@
+            {
+                return;
+            }
+
+            var parameters = method.Parameters;
+
+            // If the chosen overload has a params object[] that is being passed an array of
+            // reference types then an actual array must be being passed instead of passing one
+            // argument at a time because we know the set of overloads and how overload resolution
+            // will work.  This array may have been created somewhere we can't analyze, so never
+            // squiggle in this case.
+            if (ArgsIsArrayOfReferenceTypes(context.SemanticModel, arguments, parameters, syntaxFacts))
+            {
+                return;
+            }
+
+            var formatStringLiteralExpressionSyntax = TryGetFormatStringLiteralExpressionSyntax(
+                context.SemanticModel,
+                arguments,
+                parameters,
+                syntaxFacts);
+
+            if (formatStringLiteralExpressionSyntax == null)
+            { 
+                return;
+            }
+
+            if (parameters.Length == 0)
+            {
+                return;
+            }
+
+            var hasIFormatProvider = parameters[0].Type.Equals(formatProviderType);
+
+            // We know the format string parameter exists so numberOfArguments is at least one,
+            // and at least 2 if there is also an IFormatProvider.  The result can be zero if
+            // calling string.Format("String only with no placeholders"), where there is an
+            // empty params array.
+            var numberOfPlaceholderArguments = numberOfArguments - (hasIFormatProvider ? 2 : 1);
+
+            var formatString = formatStringLiteralExpressionSyntax.ToString();
+
+            if (FormatCallWorksAtRuntime(formatString, numberOfPlaceholderArguments))
+            {
+                return;
+            }
+
+            ValidateAndReportDiagnostic(context, numberOfPlaceholderArguments,
+                formatString, formatStringLiteralExpressionSyntax.SpanStart);
+        }
+
+        private bool IsValidFormatMethod(ISyntaxFactsService syntaxFacts, SyntaxNode expression)
+        {
+            // When calling string.Format(...), the expression will be MemberAccessExpressionSyntax
+            if (syntaxFacts.IsSimpleMemberAccessExpression(expression))
+            {
+                var nameOfMemberAccessExpression = syntaxFacts.GetNameOfMemberAccessExpression(expression);
+                return !syntaxFacts.IsGenericName(nameOfMemberAccessExpression)
+                    && syntaxFacts.GetIdentifierOfSimpleName(nameOfMemberAccessExpression).ValueText
+                    == (nameof(string.Format));
+            }
+
+            // When using static System.String and calling Format(...), the expression will be
+            // IdentifierNameSyntax
+            if (syntaxFacts.IsIdentifierName(expression))
+            {
+                return syntaxFacts.GetIdentifierOfSimpleName(expression).ValueText
```

Comment:
Won't this trigger if I have a method called `Format`? I think we need to do a semantic lookup here and see if the method symbol is a string.Format.